xla/hlo/transforms/collectives/all_gather_code_motion.h

```diff
@@ -0,0 +1,121 @@
+/* Copyright 2025 The OpenXLA Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef ALL_GATHER_CODE_MOTION_H
+#define ALL_GATHER_CODE_MOTION_H
+
+#include "xla/hlo/pass/hlo_pass_interface.h"
+
+namespace xla {
+
+// AllGatherCodeMotion optimizes HLO modules by hoisting all-gather operations
+// on while loop parameters out of the loop body into the parent computation.
+// This transformation modifies buffer shapes in-place and adjusts the while
+// loop condition and body computations accordingly. The pass also accepts a
+// replica group that forces the pass to only code motion that replica group.
+//
+// The pass handles code motion and shape adjustments for:
+// - All-gather operations and their corresponding convert ops
+// - Optimization barriers referencing the modified shapes
+// - Tuples and get-tuple-element operations that contain or extract the
+// modified values
+//
+// Example transformation:
+// clang-format off
+// while_body {
+//   body_param = (f32[128,256], f32[128,256]) parameter(0)
+//   p0 = f32[128,256] get-tuple-element(body_param), index=0
+//   p1 = f32[128,256] get-tuple-element(body_param), index=1
+//   ag0 = f32[512,256] all-gather(p0), dimensions={0}, replica_groups={{0,1,2,3}}
```

Comment:
Why is it correct to hoist this AG? Doesn't it depend on the prev iterations compute?